### ISSUE TITLE
Kubectl top no doesn't wotk when using HPA without heapster

### ISSUE THREAD

**Issue by yastij** [kubernetes/kubernetes#59438]
Currently 

```
kubectl top no
```
Doesn't work when using HPA without heapster, there's some files to change under `utilmetrics/` 

we might want this for 1.10

/assign
/sig cli

**Comment by yastij** [comment]
cc @DirectXMan12

**Comment by DirectXMan12** [comment]
can you clarify what you mean?  #56206 should have solved this issue for 1.10.

**Comment by yastij** [comment]
When using metrics-server instead of heapster top no still looks for heapster

**Comment by DirectXMan12** [comment]
Have you registered the metrics API with your API server?  Can you provide the output of `kubectl top nodes --v=10`?  Are you certain you have a fresh discovery cache?  The code is there to pull from metrics server, when it's available.

**Comment by yastij** [comment]
Yup we 've got the whole setting up and running, I'll send a reproducer next week

**Comment by pawankkamboj** [comment]
We are facing the same issue with 1.9.3 version, here is output from top nodes
```
[root@docker01 metrics-server]# kubectl top nodes --v=10
I0221 12:00:49.725780   30017 loader.go:354] Config loaded from file /root/.kube/config
I0221 12:00:49.728051   30017 cached_discovery.go:118] returning cached discovery info from /root/.kube/cache/discovery/192.168.61.6_5443/servergroups.json
I0221 12:00:49.729254   30017 round_trippers.go:398] curl -k -v -XGET  -H "Accept: application/json, */*" -H "User-Agent: kubectl/v1.6.6 (linux/amd64) kubernetes/7fa1c17" https://192.168.61.6:5443/api/v1/namespaces/kube-system/services/http:heapster:/proxy/apis/metrics/v1alpha1/nodes?labelSelector=
I0221 12:00:49.791600   30017 round_trippers.go:417] GET https://192.168.61.6:5443/api/v1/namespaces/kube-system/services/http:heapster:/proxy/apis/metrics/v1alpha1/nodes?labelSelector= 404 Not Found in 62 milliseconds
I0221 12:00:49.791640   30017 round_trippers.go:423] Response Headers:
I0221 12:00:49.791652   30017 round_trippers.go:426]     Content-Length: 192
I0221 12:00:49.791663   30017 round_trippers.go:426]     Date: Wed, 21 Feb 2018 06:30:49 GMT
I0221 12:00:49.791674   30017 round_trippers.go:426]     Content-Type: application/json
I0221 12:00:49.791742   30017 request.go:991] Response Body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"services \"heapster\" not found","reason":"NotFound","details":{"name":"heapster","kind":"services"},"code":404}
I0221 12:00:49.791974   30017 helpers.go:206] server response object: [{
  "metadata": {},
  "status": "Failure",
  "message": "the server could not find the requested resource (get services http:heapster:)",
  "reason": "NotFound",
  "details": {
    "name": "http:heapster:",
    "kind": "services",
    "causes": [
      {
        "reason": "UnexpectedServerResponse",
        "message": "unknown"
      }
    ]
  },
  "code": 404
}]
F0221 12:00:49.792072   30017 helpers.go:119] Error from server (NotFound): the server could not find the requested resource (get services http:heapster:)
```

Hete metrics server info
```
[root@docker01 metrics-server]# kubectl cluster-info
Kubernetes master is running at https://192.168.61.6:5443
CoreDNS is running at https://192.168.61.6:5443/api/v1/proxy/namespaces/kube-system/services/kube-dns
Metrics-server is running at https://192.168.61.6:5443/api/v1/proxy/namespaces/kube-system/services/metrics-server

To further debug and diagnose cluster problems, use 'kubectl cluster-info dump'.
[root@docker01 metrics-server]# kubectl get --raw "/apis/metrics.k8s.io/v1beta1/"
{"kind":"APIResourceList","apiVersion":"v1","groupVersion":"metrics.k8s.io/v1beta1","resources":[{"name":"nodes","singularName":"","namespaced":false,"kind":"NodeMetrics","verbs":["get","list"]},{"name":"pods","singularName":"","namespaced":true,"kind":"PodMetrics","verbs":["get","list"]}]}
[root@docker01 metrics-server]#
```

**Comment by DirectXMan12** [comment]
1.9.3 will have the issue, since the fix merged for 1.10

**Comment by yastij** [comment]
@DirectXMan12 - closing this one, since this is landing in 1.10.

**Comment by george-angel** [comment]
I'm getting this exact same problem:

```
kubernetes-manifests(master) 12:20:05 $ kubectl --context=exp-1 top node
Error from server (NotFound): the server could not find the requested resource (get services http:heapster:)
kubernetes-manifests(master) 12:20:31 $ kubectl --context=exp-1 version
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.4", GitCommit:"5ca598b4ba5abb89bb773071ce452e33fb66339d", GitTreeState:"clean", BuildDate:"2018-06-06T08:13:03Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.0", GitCommit:"91e7b4fd31fcd3d5f436da26c980becec37ceefe", GitTreeState:"clean", BuildDate:"2018-06-27T20:08:34Z", GoVersion:"go1.10.2", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by george-angel** [comment]
No, I'm sorry, that was my fault, I had to fix netpols to allow apiserver to access metrics-server pod.

**Comment by HaymonEdmur** [comment]
$ kubectl top pod boston-pod
Error from server (NotFound): the server could not find the requested resource (get services http:heapster:)

Is there any other way to get cpu/mem usage for a pod?

**Comment by mariojacobo** [comment]
+1 

"message": "the server is currently unable to handle the request (get services http:heapster:)",
  "reason": "ServiceUnavailable",
  "details": {
    "name": "http:heapster:",
    "kind": "services",
    "causes": [
      {
        "reason": "UnexpectedServerResponse",
        "message": "Error: 'dial tcp 10.48.0.70:8082: getsockopt: connection timed out'\nTrying to reach: 'http://10.48.0.70:8082/apis/metrics/v1alpha1/namespaces/default/pods?labelSelector='"

We have both heapster and metrics-server, so not really sure which one should fetch the information. 

service/heapster                                              ClusterIP      10.0.20.249   <none>           80/TCP                       7d
service/kube-dns                                              ClusterIP      10.0.16.10    <none>           53/UDP,53/TCP                7d
service/kubernetes-dashboard                                  ClusterIP      10.0.22.112   <none>           80/TCP                       7d
service/metrics-server                                        ClusterIP      10.0.26.120   <none>           443/TCP                      5d

**Comment by pulpbill** [comment]
Same, I have kube-state-metrics on the cluster, can't I get top from it? 

`I1214 09:38:35.675417   23770 round_trippers.go:442] Response Headers:
I1214 09:38:35.675449   23770 round_trippers.go:445]     Content-Type: application/json
I1214 09:38:35.675478   23770 round_trippers.go:445]     Content-Length: 192
I1214 09:38:35.675495   23770 round_trippers.go:445]     Date: Fri, 14 Dec 2018 12:38:35 GMT
I1214 09:38:35.675591   23770 request.go:836] Response Body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"services \"heapster\" not found","reason":"NotFound","details":{"name":"heapster","kind":"services"},"code":404}
I1214 09:38:35.676064   23770 helpers.go:207] server response object: [{
  "metadata": {},
  "status": "Failure",
  "message": "the server could not find the requested resource (get services http:heapster:)",
  "reason": "NotFound",
  "details": {
    "name": "http:heapster:",
    "kind": "services",
    "causes": [
      {
        "reason": "UnexpectedServerResponse",
        "message": "unknown"
      }
    ]
  },
  "code": 404
}]
F1214 09:38:35.676158   23770 helpers.go:120] Error from server (NotFound): the server could not find the requested resource (get services http:heapster:)
`

**Comment by zedtux** [comment]
I have a k8s 1.13 cluster running and have the same error message than @HaymonEdmur from [his comment](
https://github.com/kubernetes/kubernetes/issues/59438#issuecomment-410018142).

I've checked the presence of `metrics.k8s.io` as told by @DirectXMan12 and it was missing.
I have installed it from the [kubernetes-incubator/metrics-server](https://github.com/kubernetes-incubator/metrics-server) project with the `kubectl create -f deploy/1.8+/` command.
Now the `metrics.k8s.io` api service is present from the output of the `kubectl get apiservices` command.

Now when I'm running the top command, I have the following :

```
$ kubectl top nodes
error: metrics not available yet

$ kubectl top pods
```

The `kubectl top nodes` returns the error message directly, it sad by okay.
The `kubectl top pods` returns nothing and that's really surprising me as I have running pods.

Is there anything I'm doing wrong here?
